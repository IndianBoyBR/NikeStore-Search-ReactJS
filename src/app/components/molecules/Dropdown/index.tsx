import ThemeProviderWrapper from 'app/styles/themes/themeProvider';
import React, { memo } from 'react';
import {
  Container,
  Details,
  FieldSet,
  Input,
  Label,
  Legend,
  Options,
  Summary,
  Title,
} from './styles';

export interface DropdownOption {
  title: string;
  key: string | number;
}

export interface DropdownProps {
  label: string;
  options: DropdownOption[];
  setOptionSelected: React.Dispatch<React.SetStateAction<DropdownOption>>;
  optionSelectedState: DropdownOption;
}

export const Dropdown = memo(
  ({
    label,
    options,
    optionSelectedState,
    setOptionSelected,
  }: DropdownProps) => {
    const ref = React.useRef<HTMLDetailsElement>(null);

    const handleClick = (optionsSelected: DropdownOption) => {
      setOptionSelected(optionsSelected);
      closeSummary();
    };

    const closeSummary = () => {
      ref.current?.removeAttribute('open');
    };

    return (
      <ThemeProviderWrapper>
        <Container data-testid="dropdown">
          <FieldSet>
            <Legend>{optionSelectedState?.title ?? label}</Legend>
            <Details ref={ref} role="details">
              <Summary>{optionSelectedState?.title ?? label}</Summary>
              <Options>
                {options.map((item, index) => {
                  return (
                    <Label key={`${item.key}${item.title}`}>
                      <Input
                        defaultChecked={item.key === optionSelectedState?.key}
                        onClick={() => handleClick(item)}
                        type="radio"
                        name="radio"
                      />
                      <Title>{item.title}</Title>
                    </Label>
                  );
                })}
              </Options>
            </Details>
          </FieldSet>
        </Container>
      </ThemeProviderWrapper>
    );
  },
);
