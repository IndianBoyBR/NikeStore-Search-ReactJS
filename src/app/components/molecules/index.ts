export * from '../organisms/GoogleMapsDirectionsModal';
export * from './Dropdown';
export * from './LinksDropdown';
export * from './NewsLetter';
export * from './ShoppingStoreCard';
