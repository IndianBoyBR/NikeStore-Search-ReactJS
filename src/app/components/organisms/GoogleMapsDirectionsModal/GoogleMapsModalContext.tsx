import React from 'react';
import { ShoppingStore } from 'types';
import { LocationOrigin } from 'types/location-origin';

const GoogleMapsModalContext = React.createContext({
  modalData: {
    showModal: false,
    shoppingStoreSelected: {} as ShoppingStore,
    locationOrigin: {} as LocationOrigin,
  },
  setModalDataIntoContext: (
    _showModal: boolean,
    _shoppingStoreSelected: ShoppingStore,
    _origin: LocationOrigin,
  ) => {},
});

GoogleMapsModalContext.displayName = 'GoogleMapsModalContext';

export { GoogleMapsModalContext };
