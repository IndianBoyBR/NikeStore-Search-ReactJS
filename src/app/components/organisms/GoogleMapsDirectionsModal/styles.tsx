import styled, { css } from 'styled-components/macro';
import { ConvertHexToRGBA } from 'utils/helpers';

export const Container = styled.div`
  position: fixed;
  z-index: 999;
  left: 5%;
  top: 10%;
  width: 90%;
  height: 80%;
  overflow: auto;
  box-shadow: 0 0 4rem rgb(0 0 0 / 10%);
  border-radius: 0.625rem;
`;

export const CloseButton = styled.button`
  ${props => css`
    background-color: ${props.theme.palette.primary.default};
    background-color: ${props.theme.palette.secondary.default};
    box-shadow: 0 0.2rem 0.5rem
      ${ConvertHexToRGBA(props.theme.palette.primary.default, 0.2)};
  `}

  border-radius: 50%;
  cursor: pointer;
  outline: none;
  text-align: center;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border: none;
  padding: 1rem;
  z-index: 999;
  position: fixed;
  bottom: 0.4rem;
  right: 5%;
`;
