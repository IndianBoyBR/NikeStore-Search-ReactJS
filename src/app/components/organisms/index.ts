export * from './Footer';
export * from './Header';
export * from './NewsSlider';
export * from './ShoppingStoresList';
export * from './TopHeader';
