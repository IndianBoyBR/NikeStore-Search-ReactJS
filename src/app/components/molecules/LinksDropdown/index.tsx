import ThemeProviderWrapper from 'app/styles/themes/themeProvider';
import { memo } from 'react';
import { Details, Icon, NavLinkStyled, Options, Summary } from './styles';

export interface Link {
  label: string;
  url: string;
}

interface Props {
  title: string;
  links?: Link[];
}

export const LinksDropdown = memo(({ title, links }: Props) => {
  return (
    <ThemeProviderWrapper>
      <Details data-testid="linksDropdown">
        <Summary>
          {title}
          <Icon />
        </Summary>
        <Options>
          {links?.map((link, index) => (
            <NavLinkStyled key={`${link.url}${link.label}`} to={link.url}>
              {link.label}
            </NavLinkStyled>
          ))}
        </Options>
      </Details>
    </ThemeProviderWrapper>
  );
});
