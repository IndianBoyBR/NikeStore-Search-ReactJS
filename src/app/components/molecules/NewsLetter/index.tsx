import { Button } from 'app/components/atoms';
import Nike from 'app/components/atoms/svgs/logos/Nike';
import ThemeProviderWrapper from 'app/styles/themes/themeProvider';
import { memo } from 'react';
import { useTranslation } from 'react-i18next';
import { messages } from './messages';
import { Content, Message, Title } from './styles';

interface Props {}

export const NewsLetter = memo((props: Props) => {
  const { t } = useTranslation();

  return (
    <ThemeProviderWrapper>
      <Content data-testid="newsLetter">
        <Nike style={{ marginBottom: '1.5rem' }} />
        <Title>{t(messages.i18nTitle())}</Title>
        <Message>{t(messages.i18nMessage())}</Message>
        <Button label={t(messages.i18nRegisterButtonLabel())}></Button>
      </Content>
    </ThemeProviderWrapper>
  );
});
