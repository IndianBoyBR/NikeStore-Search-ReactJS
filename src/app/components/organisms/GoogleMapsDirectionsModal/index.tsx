import CloseIcon from 'app/components/atoms/svgs/icons/Close';
import { GoogleMapsDirections } from 'app/components/organisms/GoogleMapsDirections';
import { GoogleMapsModalContext } from 'app/components/organisms/GoogleMapsDirectionsModal/GoogleMapsModalContext';
import { memo } from 'react';
import { CloseButton, Container } from './styles';

export const GoogleMapsDirectionsModal = memo(() => {
  return (
    <GoogleMapsModalContext.Consumer>
      {({ modalData, setModalDataIntoContext }) => {
        if (!modalData?.showModal) {
          return null;
        }

        return (
          <>
            <Container>
              <GoogleMapsDirections
                origin={{
                  lat: modalData.locationOrigin.lat,
                  lng: modalData.locationOrigin.lng,
                }}
                destination={{
                  lat: modalData.shoppingStoreSelected.address.lat,
                  lng: modalData.shoppingStoreSelected.address.lng,
                }}
              />
            </Container>
            <CloseButton
              onClick={() =>
                setModalDataIntoContext(
                  false,
                  {
                    id: 0,
                    placeName: '',
                    address: {
                      lat: 0,
                      lng: 0,
                      location: '',
                    },
                    availability: 0,
                    workingHours: '',
                  },
                  {
                    lat: 0,
                    lng: 0,
                    placeName: '',
                  },
                )
              }
            >
              <CloseIcon />
            </CloseButton>
          </>
        );
      }}
    </GoogleMapsModalContext.Consumer>
  );
});
