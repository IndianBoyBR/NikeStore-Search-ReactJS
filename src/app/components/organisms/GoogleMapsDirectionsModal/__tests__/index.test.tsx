import { Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

import { GoogleMapsModalContext } from 'app/components/organisms/GoogleMapsDirectionsModal/GoogleMapsModalContext';

import ThemeProviderWrapper from 'app/styles/themes/themeProvider';

import { GoogleMapsDirectionsModal } from 'app/components/molecules';
import { ShoppingStoresMockData } from 'mocks/nike-stores/shopping-stores-data';
import { useMemo, useState } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { configureAppStore } from 'store/configureStore';
import { ShoppingStore } from 'types';

import { initialize as initializeGoogleMapsJestMocks } from '@googlemaps/jest-mocks';
import { LocationOrigin } from 'types/location-origin';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: str => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));

beforeEach(() => {
  initializeGoogleMapsJestMocks();

  global.window.google.maps = {
    ...global.window.google.maps,
    TravelMode: {
      DRIVING: 'DRIVING',
    },
  } as any;
});

const ComponentComponentWithParentAndProviders = (store: Store) => {
  const [modalData, setModalData] = useState<{
    showModal: boolean;
    shoppingStoreSelected: ShoppingStore;
    locationOrigin: LocationOrigin;
  }>({
    showModal: true,
    shoppingStoreSelected: ShoppingStoresMockData[0],
    locationOrigin: {
      placeName: '',
      lat: 0,
      lng: 0,
    },
  });

  const setModalDataIntoContext = (
    showModal: boolean,
    shoppingStoreSelected: ShoppingStore,
    locationOrigin: any,
  ) => {
    setModalData({
      showModal,
      shoppingStoreSelected,
      locationOrigin,
    });
  };

  const modalContextMemo = useMemo(
    () => ({ modalData, setModalDataIntoContext }),
    [modalData],
  );

  return (
    <Provider store={store}>
      <Router>
        <ThemeProviderWrapper>
          <GoogleMapsModalContext.Provider value={modalContextMemo}>
            <GoogleMapsDirectionsModal />
          </GoogleMapsModalContext.Provider>
        </ThemeProviderWrapper>
      </Router>
    </Provider>
  );
};

describe('<GoogleMapsDirectionsModal  />', () => {
  let store: ReturnType<typeof configureAppStore>;

  beforeEach(() => {
    store = configureAppStore();
  });

  it('should match snapshot', () => {
    const ParentComponent = () =>
      ComponentComponentWithParentAndProviders(store);
    const { container } = render(<ParentComponent />);

    const googleMaps = container.querySelector<HTMLDivElement>('#google-map');

    expect(googleMaps).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
