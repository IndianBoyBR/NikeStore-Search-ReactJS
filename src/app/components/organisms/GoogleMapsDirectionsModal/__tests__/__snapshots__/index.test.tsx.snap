// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoogleMapsDirectionsModal  /> should match snapshot 1`] = `
.c0 {
  position: fixed;
  z-index: 999;
  left: 5%;
  top: 10%;
  width: 90%;
  height: 80%;
  overflow: auto;
  box-shadow: 0 0 4rem rgb(0 0 0 / 10%);
  border-radius: 0.625rem;
}

<div
  class="c0"
>
  <div
    id="google-map"
    style="border-radius: 0.625rem; height: 100%; width: 100%;"
  />
</div>
`;
